### PR TITLE
Codegen Index Scan

### PR DESCRIPTION
**Pull request by huangminchn**
This PR adds support for index scan code generation. (require fixes from #936 and #937 )

### CODE REVIEW

**Comment by apavlo** [comment]
@huangminchn Can you fix this so that we can merge it?

**Comment by huangminchn** [comment]
@apavlo Sure. All it needs is a codegen cache patch. I will fix it ASAP.

**Comment by apavlo** [comment]
@pmenon @seojungmin This is an important feature that we need to merge soon. Can you look this over?

File: src/codegen/operator/index_scan_translator.cpp [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+        default: {
+          throw new Exception("Type" +
+                              peloton::TypeIdToString(ai->type.type_id) +
+                              " is not supported in codegen yet");
+        }
+      }
+    }
+  }
+}
+}
```
Comment (seojungmin) [inline src/codegen/operator/index_scan_translator.cpp]: Formatting - add a new line between 341 and 342, then add `// namespace codegen` and `// namespace peloton` on the last two lines.

File: src/codegen/proxy/index_proxy.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#include "codegen/proxy/index_proxy.h"
+
+namespace peloton {
+namespace codegen {
+
+DEFINE_TYPE(Index, "index::Index", MEMBER(opaque));
+// DEFINE_METHOD(peloton::index, Index, CodeGenScan);
```
Comment (seojungmin) [inline src/codegen/proxy/index_proxy.cpp]: Remove this commented line.

File: src/codegen/proxy/index_scan_iterator_proxy.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+DEFINE_METHOD(peloton::codegen::util, IndexScanIterator, GetTileGroupId);
+DEFINE_METHOD(peloton::codegen::util, IndexScanIterator, RowOffsetInResult);
+DEFINE_METHOD(peloton::codegen::util, IndexScanIterator, GetResultSize);
+DEFINE_METHOD(peloton::codegen::util, IndexScanIterator, GetTileGroupOffset);
+DEFINE_METHOD(peloton::codegen::util, IndexScanIterator, UpdateTupleWithInteger);
+DEFINE_METHOD(peloton::codegen::util, IndexScanIterator, UpdateTupleWithBigInteger);
+DEFINE_METHOD(peloton::codegen::util, IndexScanIterator, UpdateTupleWithDouble);
+DEFINE_METHOD(peloton::codegen::util, IndexScanIterator, UpdateTupleWithVarchar);
+DEFINE_METHOD(peloton::codegen::util, IndexScanIterator, UpdateTupleWithBoolean);
+}
```
Comment (seojungmin) [inline src/codegen/proxy/index_scan_iterator_proxy.cpp]: Formatting - add a new lines and add `// namespace codegen` and `// namespace peloton` on the last two lines.

File: src/codegen/proxy/item_pointer_proxy.cpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+// Copyright (c) 2015-2017, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "codegen/proxy/item_pointer_proxy.h"
+
+namespace peloton {
+namespace codegen {
+DEFINE_TYPE(ItemPointer, "ItemPointer", MEMBER(opaque));
+}
```
Comment (seojungmin) [inline src/codegen/proxy/item_pointer_proxy.cpp]: Formatting - add `// namespace codegen` and `// namespace peloton` on the last two lines.

File: src/codegen/runtime_functions.cpp [hunk 5]
```diff
@@ -71,6 +73,12 @@ storage::TileGroup *RuntimeFunctions::GetTileGroup(storage::DataTable *table,
   return tile_group.get();
 }
 
+storage::TileGroup *RuntimeFunctions::GetTileGroupByGlobalId(
```
Comment (seojungmin) [inline src/codegen/runtime_functions.cpp]: Change the name `GetTileGroupByGlobalId` to `GetTileGroupById` to make the API names consistent.

File: src/codegen/runtime_functions.cpp [hunk 6]
```diff
@@ -129,5 +137,20 @@ void RuntimeFunctions::ThrowOverflowException() {
   throw std::overflow_error("ERROR: overflow");
 }
 
+util::IndexScanIterator *RuntimeFunctions::GetIterator(index::Index *index,
```
Comment (seojungmin) [inline src/codegen/runtime_functions.cpp]: Can we have some helper for the index scan translator, such as `sorter.cpp` and `inserter.cpp`?  Having this and `DeleteIterator()` without some _holder_ object would be dangerous and makes the code less readable.  It may cause memory leaks, if there is any exception thrown between them, and it is hard to catch them later.

i.e. make a helper object and move these into the helper.

File: src/codegen/translator_factory.cpp [hunk 7]
```diff
@@ -33,6 +33,8 @@
 #include "codegen/operator/table_scan_translator.h"
 #include "codegen/operator/update_translator.h"
 #include "expression/aggregate_expression.h"
+#include "codegen/operator/index_scan_translator.h"
+#include "codegen/expression/tuple_value_translator.h"
```
Comment (seojungmin) [inline src/codegen/translator_factory.cpp]: Check the rule for include files's order.

File: src/codegen/util/index_scan_iterator.cpp [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+  if (is_point_query_) {
+    index_->ScanKey(point_key_p_, result_);
+  } else if (is_full_scan_) {
+    index_->CodeGenFullScan(result_);
+  } else {
+    index_->CodeGenRangeScan(low_key_p_, high_key_p_, result_);
+  }
+  LOG_TRACE("result size = %lu\n", result_.size());
+
+  // TODO: codes below might be useful for optimization
```
Comment (seojungmin) [inline src/codegen/util/index_scan_iterator.cpp]: State why they might be useful  for optimization.

File: src/codegen/util/index_scan_iterator.cpp [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+  for (unsigned int i = 0; i < columns.size(); i++) {
+    if (strcmp(columns[i].GetName().c_str(), attribute_name) == 0) {
+      update_tuple->SetValue(
+          i, peloton::type::ValueFactory::GetBooleanValue(value));
+      break;
+    }
+  }
+}
+}
+}
```
Comment (seojungmin) [inline src/codegen/util/index_scan_iterator.cpp]: Formatting

File: src/include/codegen/index.h [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace codegen {
+class Index {
+ public:
+  Index(index::Index &index);
+
+ private:
+  index::Index &index_;
+};
+}
```
Comment (seojungmin) [inline src/include/codegen/index.h]: Formatting

File: src/include/codegen/operator/index_scan_translator.h [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  // The scan
+  const planner::IndexScanPlan &index_scan_;
+
+  // The ID of the selection vector in runtime state
+  RuntimeState::StateID selection_vector_id_;
+
+  // The code-generating table instance
+  //  codegen::ArtIndex index_;
+};
+}
```
Comment (seojungmin) [inline src/include/codegen/operator/index_scan_translator.h]: Formatting

File: src/include/codegen/proxy/index_proxy.h [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+namespace codegen {
+PROXY(Index) {
+  DECLARE_MEMBER(0, char[sizeof(index::Index)], opaque);
+  DECLARE_TYPE;
+
+  //  DECLARE_METHOD(CodeGenScan);
+};
+
+TYPE_BUILDER(Index, index::Index);
+}
```
Comment (seojungmin) [inline src/include/codegen/proxy/index_proxy.h]: Formatting

File: src/include/codegen/proxy/index_proxy.h [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+#include "codegen/proxy/type_builder.h"
+#include "index/index.h"
+
+namespace peloton {
+namespace codegen {
+PROXY(Index) {
+  DECLARE_MEMBER(0, char[sizeof(index::Index)], opaque);
+  DECLARE_TYPE;
+
+  //  DECLARE_METHOD(CodeGenScan);
```
Comment (seojungmin) [inline src/include/codegen/proxy/index_proxy.h]: Remove this commented line.

File: src/include/codegen/proxy/index_scan_iterator_proxy.h [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+  DECLARE_METHOD(GetTileGroupOffset);
+  DECLARE_METHOD(UpdateTupleWithInteger);
+  DECLARE_METHOD(UpdateTupleWithBigInteger);
+  DECLARE_METHOD(UpdateTupleWithDouble);
+  DECLARE_METHOD(UpdateTupleWithVarchar);
+  DECLARE_METHOD(UpdateTupleWithBoolean);
+};
+
+TYPE_BUILDER(IndexScanIterator, util::IndexScanIterator);
+}
```
Comment (seojungmin) [inline src/include/codegen/proxy/index_scan_iterator_proxy.h]: Formatting

File: src/include/codegen/proxy/item_pointer_proxy.h [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace peloton {
+namespace codegen {
+PROXY(ItemPointer) {
+  DECLARE_MEMBER(0, char[sizeof(ItemPointer)], opaque);
+  DECLARE_TYPE;
+};
+
+TYPE_BUILDER(ItemPointer, ItemPointer);
+}
```
Comment (seojungmin) [inline src/include/codegen/proxy/item_pointer_proxy.h]: Formatting

File: src/include/codegen/runtime_functions.h [hunk 16]
```diff
@@ -23,13 +23,25 @@ class TileGroup;
...
+}  // namespace expression
+
+namespace index {
+class Index;
+}
+
+namespace codegen {
+namespace util {
+class IndexScanIterator;
+}
```
Comment (seojungmin) [inline src/include/codegen/runtime_functions.h]: Formatting

File: src/include/codegen/util/index_scan_iterator.h [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+                                 char *attribute_name, bool is_lower_key);
+  void UpdateTupleWithDouble(double value, int attribute_id,
+                             char *attribute_name, bool is_lower_key);
+  void UpdateTupleWithVarchar(char *value, int attribute_id,
+                              char *attribute_name, bool is_lower_key);
+  void UpdateTupleWithBoolean(bool value, int attribute_id,
+                              char *attribute_name, bool is_lower_key);
+};
+}
+}
```
Comment (seojungmin) [inline src/include/codegen/util/index_scan_iterator.h]: Formatting

File: src/codegen/compilation_context.cpp [hunk 18]
```diff
@@ -81,7 +81,6 @@ void CompilationContext::GeneratePlan(QueryCompiler::CompileStats *stats) {
 
   // First we prepare the translators for all the operators in the tree
   Prepare(query_.GetPlan(), main_pipeline_);
-
```
Comment (pmenon) [inline src/codegen/compilation_context.cpp]: Remove.

File: src/codegen/index.cpp [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+#include "catalog/schema.h"
+#include "codegen/proxy/data_table_proxy.h"
+#include "codegen/lang/loop.h"
+#include "codegen/proxy/runtime_functions_proxy.h"
+#include "index/index.h"
+
+namespace peloton {
+namespace codegen {
+
+Index::Index(index::Index &index) : index_(index) {}
```
Comment (pmenon) [inline src/codegen/index.cpp]: I don't think this class is used. You can remove it.

File: src/codegen/operator/index_scan_translator.cpp [hunk 20]
```diff
@@ -0,0 +1,343 @@
...
+
+  // The restriction, if one exists
+  const auto *predicate = index_scan_.GetPredicate();
+
+  if (predicate != nullptr) {
+    // If there is a predicate, prepare a translator for it
+    context.Prepare(*predicate);
+
+    // If the scan's predicate is SIMDable, install a boundary at the output
+    if (predicate->IsSIMDable()) {
```
Comment (pmenon) [inline src/codegen/operator/index_scan_translator.cpp]: Remove this check.

File: src/codegen/operator/index_scan_translator.cpp [hunk 21]
```diff
@@ -0,0 +1,343 @@
...
+    llvm::Value *txn = codegen.Call(ExecutorContextProxy::GetTransaction,
+                                    {executor_context_ptr});
+    llvm::Value *raw_sel_vec = sel_vec.GetVectorPtr();
+    // Invoke TransactionRuntime::PerformRead(...)
+    llvm::Value *out_idx =
+        codegen.Call(TransactionRuntimeProxy::PerformVectorizedRead,
+                     {txn, tile_group_ptr, tile_group_offset,
+                      codegen->CreateAdd(tile_group_offset, codegen.Const32(1)),
+                      raw_sel_vec});
+    sel_vec.SetNumElements(out_idx);
```
Comment (pmenon) [inline src/codegen/operator/index_scan_translator.cpp]: A lot of this code is duplicated from `codegen::TileGroup` and `codegen::Table`. Can you try to integrate it more cleanly?

File: src/codegen/operator/index_scan_translator.cpp [hunk 22]
```diff
@@ -0,0 +1,343 @@
...
+      for (const auto *ai : used_attributes) {
+        attribute_accessors.emplace_back(tile_group_access, ai);
+      }
+      for (uint32_t i = 0; i < attribute_accessors.size(); i++) {
+        auto &accessor = attribute_accessors[i];
+        batch.AddAttribute(accessor.GetAttributeRef(), &accessor);
+      }
+
+      // Iterate over the batch using a scalar loop
+      batch.Iterate(codegen, [&](RowBatch::Row &row) {
```
Comment (pmenon) [inline src/codegen/operator/index_scan_translator.cpp]: Can you pull this out into a separate function?

File: src/codegen/query_compiler.cpp [hunk 23]
```diff
@@ -10,6 +10,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "planner/index_scan_plan.h"
```
Comment (pmenon) [inline src/codegen/query_compiler.cpp]: Remove.

File: src/codegen/query_compiler.cpp [hunk 24]
```diff
@@ -88,6 +90,11 @@ bool QueryCompiler::IsSupported(const planner::AbstractPlan &plan) {
       pred = scan_plan.GetPredicate();
       break;
     }
+    case PlanNodeType::INDEXSCAN: {
```
Comment (pmenon) [inline src/codegen/query_compiler.cpp]: Merge this case with SeqScanPlan, cast to `AbstractScanPlan` and check predicate.

File: src/codegen/util/index_scan_iterator.cpp [hunk 25]
```diff
@@ -0,0 +1,217 @@
...
+  //        begin = iter;
+  //      }
+  //      ++iter;
+  //    }
+  //    // don't forget the last one
+  //    result_metadata_.push_back(previous_tile_group_id);
+  //    result_metadata_.push_back(begin);
+  //    result_metadata_.push_back(iter);
+  //    distinct_tile_group_num_++;
+  //  }
```
Comment (pmenon) [inline src/codegen/util/index_scan_iterator.cpp]: Remove commented code. If you leave `TODO`, state in plain english what optimization can be made.

File: src/executor/plan_executor.cpp [hunk 26]
```diff
@@ -38,7 +38,6 @@ static void CompileAndExecutePlan(
   // Perform binding
   planner::BindingContext context;
   plan->PerformBinding(context);
-
```
Comment (pmenon) [inline src/executor/plan_executor.cpp]: Remove.

File: src/expression/abstract_expression.cpp [hunk 27]
```diff
@@ -55,6 +55,23 @@ void AbstractExpression::GetUsedAttributes(
   }
 }
 
+void AbstractExpression::GetUsedAttributesInPredicateOrder(
```
Comment (pmenon) [inline src/expression/abstract_expression.cpp]: How is this function different than `GetUsedAttributes`?

File: src/include/codegen/index.h [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+#include "codegen/codegen.h"
+#include "codegen/scan_callback.h"
+
+namespace peloton {
+namespace index {
+class Index;
+}
+
+namespace codegen {
+class Index {
```
Comment (pmenon) [inline src/include/codegen/index.h]: Remove this class altogether. It isn't used.

File: src/include/codegen/query_compiler.h [hunk 29]
```diff
@@ -66,7 +66,6 @@ class QueryCompiler {
   static bool IsExpressionSupported(const expression::AbstractExpression &plan);
 
  private:
-
```
Comment (pmenon) [inline src/include/codegen/query_compiler.h]: Remove.

File: src/type/types.cpp [hunk 30]
```diff
@@ -804,6 +804,28 @@ ExpressionType StringToExpressionType(const std::string &str) {
...
+    {
+      return true;
+    }
+    default:
+    {
+      return false;
+    }
+  }
+}
+
```
Comment (pmenon) [inline src/type/types.cpp]: Formatting.

File: src/codegen/operator/index_scan_translator.cpp [hunk 31]
```diff
@@ -0,0 +1,343 @@
...
+      bool is_lower_key = false;
+      if (comparison_type[i] == peloton::ExpressionType::COMPARE_GREATERTHAN ||
+          comparison_type[i] ==
+              peloton::ExpressionType::COMPARE_GREATERTHANOREQUALTO) {
+        is_lower_key = true;
+      }
+      llvm::Value *is_lower = codegen.ConstBool(is_lower_key);
+
+      llvm::Value *parameter_value = parameter_cache.GetValue(i).GetValue();
+      switch (ai->type.type_id) {
```
Comment (seojungmin) [inline src/codegen/operator/index_scan_translator.cpp]: Can we update the tuple value using `table_storage.cpp` so that we can remove these function calls and also the overhead setting the value through non-codegen'ed type system?  Obviously, as we talked before, using the tuple format is not the correct way to handle this and to be refactorized.  For this PR, this would be enough.

File: src/codegen/operator/index_scan_translator.cpp [hunk 32]
```diff
@@ -0,0 +1,343 @@
...
+  name.append(")");
+  return name;
+}
+
+// Index accessor
+const index::Index &IndexScanTranslator::GetIndex() const {
+  return dynamic_cast<index::Index &>(*index_scan_.GetIndex().get());
+}
+
+void IndexScanTranslator::UpdateTupleWithParameterCache(
```
Comment (seojungmin) [inline src/codegen/operator/index_scan_translator.cpp]: Name this as `SetPredicate` or something that shows what is doing in the perspective of IndexScan.  `TupleWithParameterCache` is an implementation detail.  Additionally, move the predicate check outside the function for readability.

**Comment by huangminchn** [comment]
@seojungmin @pmenon Understood. I will fix it ASAP.

File: src/codegen/runtime_functions.cpp (hunk 6 above)
Comment (huangminchn) [reply src/codegen/runtime_functions.cpp]: Hey Tony, the iterator is already a helper object; I guess what you mean is to construct the iterator with `context.GetRuntimeState().RegisterState()` function in the index_scan_translator constructor instead of calling the `GetIterator()` and `DeleteIterator()` function, right?

Comment (seojungmin) [reply src/codegen/runtime_functions.cpp]: @huangminchn Yes. The point is on the lifetime of this iterator object. The lifetime of the iterator should be the same or shorter than the one registered with the runtime state, after all.

**Comment by apavlo** [comment]
@huangminchn Any love for this PR?

**Comment by huangminchn** [comment]
@apavlo mb. I was bush recently. I will get it done today.

**Comment by huangminchn** [comment]
@pmenon The optimizer refactor pr was merged just now so I rebased; the optimizer_sql_test will fail but I couldn't figure out the reason. It seemed like the BlockNestedLoopJoinTranslator is causing the error. Do u mind taking a look at the errors from optimizer_sql_test? I will really appreciate that.

File: src/expression/abstract_expression.cpp (hunk 27 above)
Comment (huangminchn) [reply src/expression/abstract_expression.cpp]: `GetUsedAttributes` function adds attributes to a set, so when we retrieve attributes from set, the order is unpredictable.
The new parameters in the codegen cache are in predicate order; in `GetUsedAttributesInPredicateOrder` function, the attributes are pushed to a vector; therefore, I can have a correct mapping from values in new parameters to attributes in the predicate.

**Comment by pmenon** [comment]
@huangminchn The test log output doesn't provide details of the error. Can you add debug statements and see what the error is? I'm on Slack. If it's a real problem, I'll have to checkout your branch and take a look.
